### PR TITLE
Validation for OAuth2, Introduce Strategy, StrategyFactory, FileSession and RedisSession

### PR DESCRIPTION
**Pull request by udaydenduluri33**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/auth/models/index.ts [hunk 1]
```diff
@@ -1 +1,2 @@
-export { Authentication } from './authentication.class'
+export { Strategy } from './strategy.class'
+export { default as strategyFactory } from './strategyFactory.class'
```
Comment (paddybasi) [inline src/auth/models/index.ts:2]: not a big issue, but any default export becomes the named export anyway, no need to be explicit, e.g.
```
export strategyFactory from './strategyFactory.class'
```

File: src/auth/models/strategy.class.ts [hunk 2]
```diff
@@ -3,15 +3,24 @@ import * as express from 'express'
...
 import { AUTH } from '../auth.constants'
+import { OAuth2Metadata } from '../oauth2'
+import { OpenIDMetadata } from '../oidc'
+import { FileSessionMetadata, RedisSessionMetadata } from '../session/models/sessionMetadata.interface'
 
-export class Authentication extends events.EventEmitter {
-    protected readonly strategyName: string
+export abstract class Strategy extends events.EventEmitter {
+    public readonly strategyName: string
     protected readonly router = express.Router({ mergeParams: true })
```
Comment (paddybasi) [inline src/auth/models/strategy.class.ts:12]: could we just tidy this up a bit? we have 2 imports from express, 
```
import * as express
import { NextFunction, Request, RequestHandler, Response } from 'express'
```

Think I would prefer if we only import what we need, e.g.
```
import { NextFunction, Request, RequestHandler, Response, Router } from 'express'
```

File: src/auth/models/strategy.class.ts [hunk 3]
```diff
@@ -3,15 +3,24 @@ import * as express from 'express'
...
     constructor(strategyName: string) {
         super()
         this.strategyName = strategyName
     }
 
+    public abstract configure(
+        options: OAuth2Metadata | OpenIDMetadata | FileSessionMetadata | RedisSessionMetadata,
```
Comment (paddybasi) [inline src/auth/models/strategy.class.ts:19]: why does the base authentication class have session stuff? sorry not seen it before. but session should live in it's own layer/folder etc, it's independent of authentication

File: src/auth/models/strategy.class.ts [hunk 4]
```diff
@@ -41,7 +50,7 @@ export class Authentication extends events.EventEmitter {
                     console.log(`redirecting, no listener count: ${AUTH.EVENT.AUTHENTICATE_SUCCESS}`, req.session)
                     res.redirect(AUTH.ROUTE.DEFAULT_REDIRECT)
                 } else {
-                    this.emit(AUTH.EVENT.AUTHENTICATE_SUCCESS, false, req, res, next)
+                    this.emit(AUTH.EVENT.AUTHENTICATE_SUCCESS, this, false, req, res, next)
```
Comment (paddybasi) [inline src/auth/models/strategy.class.ts:53]: why do we need to now pass the instance as a reference?

File: src/auth/models/index.ts (hunk 1 above)
Comment (udaydenduluri33) [reply src/auth/models/index.ts:2]: Tried it @paddybasi. Not sure it works

File: src/auth/models/strategyFactory.class.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import oAuth2 from '../oauth2/models/oauth2.class'
+import openid from '../oidc/models/openid.class'
+import redisSession from '../session/models/redisSessionStrategy.class'
+import fileSession from '../session/models/fileSessionStrategy.class'
+import { Strategy } from './strategy.class'
+import { OAUTH2 } from '../oauth2'
+import { OIDC } from '../oidc'
+import { SESSION } from '../session/session.constants'
+
+export type StrategyType = 'oauth2' | 'oidc' | 's2s' | 'filesession' | 'redissession'
```
Comment (paddybasi) [inline src/auth/models/strategyFactory.class.ts:10]: again a 'session' isn't a strategy/authentication type, can we split this into it's own layer please

File: src/auth/models/strategy.class.ts (hunk 2 above)
Comment (udaydenduluri33) [reply src/auth/models/strategy.class.ts:12]: Agreed, done

**Pull request merged**